wallet/src/Pos/Wallet/Web/Mode.hs

```diff
@@ -7,90 +7,127 @@ module Pos.Wallet.Web.Mode
+import qualified Pos.Util.OutboundQueue           as OQ.Reader
+import           Pos.Util.TimeWarp                (CanJsonLog (..))
+import           Pos.Util.UserSecret              (HasUserSecret (..))
+import           Pos.Util.Util                    (postfixLFields)
+import           Pos.Wallet.Redirect              (MonadBlockchainInfo (..),
+                                                   MonadUpdates (..),
+                                                   applyLastUpdateWebWallet,
+                                                   blockchainSlotDurationWebWallet,
+                                                   connectedPeersWebWallet,
+                                                   localChainDifficultyWebWallet,
+                                                   networkChainDifficultyWebWallet,
+                                                   waitForUpdateWebWallet)
+import           Pos.Wallet.SscType               (WalletSscType)
+import           Pos.Wallet.Web.ClientTypes       (Addr, CHash, CId (..), cIdToAddress)
+import           Pos.Wallet.Web.Error             (WalletError (..))
+import           Pos.Wallet.Web.Sockets.ConnSet   (ConnectionsVar)
+import           Pos.Wallet.Web.State.State       (WalletState)
+import           Pos.Wallet.Web.Tracking          (MonadBListener (..), onApplyTracking,
+                                                   onRollbackTracking)
+import           Pos.WorkMode                     (RealModeContext (..))
 
 
 
 
 data WalletWebModeContext = WalletWebModeContext
     { wwmcWalletState     :: !WalletState
     , wwmcConnectionsVar  :: !ConnectionsVar
+    , wwmcHashes          :: !AddrCIdHashes
     , wwmcRealModeContext :: !(RealModeContext WalletSscType)
     }
 
+newtype AddrCIdHashes = AddrCIdHashes { unAddrCIdHashes :: (IORef (Map CHash Address)) }
+
 makeLensesWith postfixLFields ''WalletWebModeContext
 
+instance HasLens AddrCIdHashes WalletWebModeContext AddrCIdHashes where
+    lensOf = wwmcHashes_L
+
+cIdToAddress' :: (MonadWalletWebMode m) => CId w -> m Address
+cIdToAddress' = either (throwM . DecodeError) pure . cIdToAddress
+
+convertCIdTOAddr :: (MonadWalletWebMode m) => CId Addr -> m Address
+convertCIdTOAddr i@(CId id) = do
+    hmRef <- unAddrCIdHashes <$> view (lensOf @AddrCIdHashes)
+    hm <- liftIO $ readIORef hmRef
+    case id `M.lookup` hm of
+       Just addr -> return addr
+       _ -> cIdToAddress' i >>= \addr ->
+            liftIO (modifyIORef' hmRef (M.insert id addr)) $> addr
+
+convertCIdTOAddrs :: (MonadWalletWebMode m, Traversable t) => t (CId Addr) -> m (t Address)
+convertCIdTOAddrs cids = do
+    hmRef <- unAddrCIdHashes <$> view (lensOf @AddrCIdHashes)
+    hm <- liftIO $ readIORef hmRef
+    forM cids $ \i@(CId id) -> do
+        case id `M.lookup` hm of
+           Just addr -> return addr
+           _ -> cIdToAddress' i >>= \addr ->
+                    liftIO (modifyIORef' hmRef (M.insert id addr)) $> addr
```

Comment:
This will complicate semantics. We need only simple semantics for the cache. It's not atomic, so we won't have contention (and it will always take same time to update it). And we don't care much of cache not being updated eventually.

Reply:
This will complicate semantics. We need only simple semantics for the cache. It's not atomic, so we won't have contention (and it will always take same time to update it). And we don't care much of cache not being updated eventually.